README.rst
==========
syslog2IRC
==========

.. image:: https://raw.githubusercontent.com/homeworkprod/syslog2irc/main/assets/syslog2irc_logo.svg
   :alt: syslog2IRC logo
   :height: 80
   :width: 310

Forward syslog messages to IRC

For a more generic forwarder to IRC, check out `Weitersager <https://homework.nwsnet.de/releases/1cda/#weitersager>`_.


Code Status
===========

|badge_travis-ci_build|
|badge_scrutinizer-ci_coverage|
|badge_scrutinizer-ci_quality-score|
|badge_code-climate_maintainability|


.. |badge_travis-ci_build| image:: https://travis-ci.org/homeworkprod/syslog2irc.svg?branch=main
   :alt: Build Status
   :target: https://travis-ci.org/homeworkprod/syslog2irc

.. |badge_scrutinizer-ci_coverage| image:: https://scrutinizer-ci.com/g/homeworkprod/syslog2irc/badges/coverage.png?b=main
   :alt: Scrutinizer Code Coverage
   :target: https://scrutinizer-ci.com/g/homeworkprod/syslog2irc/?branch=main

.. |badge_scrutinizer-ci_quality-score| image:: https://scrutinizer-ci.com/g/homeworkprod/syslog2irc/badges/quality-score.png?b=main
   :alt: Scrutinizer Code Quality
   :target: https://scrutinizer-ci.com/g/homeworkprod/syslog2irc/?branch=main

.. |badge_code-climate_maintainability| image:: https://api.codeclimate.com/v1/badges/7fc3553f7a7cf6502560/maintainability
   :alt: Code Climate
   :target: https://codeclimate.com/github/homeworkprod/syslog2irc


Requirements
============

- Python 3.7+
- Dependencies: blinker_, irc_, rtoml_, syslogmp_

.. _blinker:  https://pythonhosted.org/blinker/
.. _irc:      https://bitbucket.org/jaraco/irc
.. _rtoml:    https://github.com/samuelcolvin/rtoml
.. _syslogmp: https://homework.nwsnet.de/releases/76d6/#syslogmp


Installation
============

syslog2IRC and its dependencies can be installed via pip_:

.. code:: sh

    $ pip install syslog2irc

To update an installed copy of syslog2IRC to the most recent release:

.. code:: sh

    $ pip install -U syslog2irc

.. _pip:      http://www.pip-installer.org/


Configuration
=============


syslog
------

Setup your ``syslog.conf`` or ``rsyslog.conf`` (commonly found in
``/etc``) to send syslog messages to syslog2IRC on the default syslog
port (514, UDP, as assigned by IANA_)::

    *.*     @host-to-send-log-messages-to-and-this-script-runs-on

Or, when syslog2IRC listens on a non-default port (here: 11514)::

    *.*     @host-to-send-log-messages-to-and-this-script-runs-on:11514

.. _IANA:     https://www.iana.org/


syslog2IRC
----------

Configuration is done as a file in TOML_ format.

A simple configuration to route from the default syslog port, 514, to a
single IRC channel looks like this:

.. code:: toml

    [irc.server]
    host = "irc.server.example"

    [irc.bot]
    nickname = "syslog"

    [irc]
    channels = [
      { name = "#syslog" },
    ]

    [routes]
    "514/udp" = [ '#syslog' ]

In a more complex setup, syslog messages could be received on multiple
ports (514/UDP and 55514/TCP in this example), with those received on
the first port being forwarded to two IRC channels, and those received
on the latter port being forwarded exclusively to the second channel.

Here is a full example of an advanced configuration with all optional
properties being specified:

.. code:: toml

    log_level = "warning"        # optional

    [irc.server]
    host = "irc.server.example"  # optional
    port = 6667                  # optional
    ssl = false                  # optional
    password = "t0ps3cr3t"       # optional
    rate_limit = 0.5             # optional; limit of messages per second

    [irc.bot]
    nickname = "syslog"
    realname = "syslog"          # optional

    [irc]
    commands = [                 # optional
      "MODE syslog +i",
    ]
    channels = [
      { name = "#examplechannel1" },
      { name = "#examplechannel2", password = "zePassword" },
    ]

    [routes]
    # routing for syslog messages from the ports on which they are
    # received to the IRC channels they should be announced on
    "514/udp" = [ '#examplechannel1', '#examplechannel2' ]
    "55514/tcp" = [ '#examplechannel2' ]

.. _TOML: https://toml.io/


IRC Dummy Mode
==============

If no value for ``irc.server.host`` is set (the property is missing or
commented out), syslog2IRC will not attempt to connect to an IRC server
and start in IRC dummy mode.

In this mode, it will still receive syslog messages, but it will write
them to STDOUT. This can be helpful during setup of syslog message
reception.

Abort execution by pressing <Control-C>.


Usage
=====

Start syslog2IRC with a configuration file:

.. code:: sh

    $ syslog2irc config.toml

Send some messages to syslog2IRC using your system's syslog message
sender tool (`logger`, in this example):

.. code:: sh

    $ logger 'Hi there!'
    $ logger -p kern.alert 'Whoa!'

Note that each message will appear twice on the console syslog2IRC was
started because the handler itself will write it there anyway (so you
have a log on what would be sent to IRC).

If receiving syslog messages works and you have been using IRC dummy
mode so far, specify an IRC server in the configuration file, then start
as above:

.. code:: sh

    $ syslog2irc config.toml

After a moment, you should see that syslog2IRC has connected to the IRC
server. The bot should then enter the channel(s) you have configured
(see Configuration_).


Custom Message Format
=====================

The application's entry point is prepared to accept a custom callable to
format messages.

Arguably the easiest way to make use of this without fiddling inside of
the installed package's sources:

- Copy the Python code from ``src/syslog2irc/formatting.py`` to a new
  file outside of the package path, e.g. ``syslog2irc-custom.py``.
- Adjust the copy of the function ``format_message`` as desired.
- Import the entry point function into the new file, then call it while
  passing the adjusted formatter function to it:

  .. code:: python

      from syslog2irc.main import main

      if __name__ == '__main__':
          main(custom_format_message=format_message)

- Run the new file in the shell:

  .. code:: sh

      $ python syslog2irc-custom.py config.toml


Further Reading
===============

For more information, see `RFC 3164`_, "The BSD syslog Protocol".

Please note that there is `RFC 5424`_, "The Syslog Protocol", which
obsoletes `RFC 3164`_. syslog2IRC, however, only implements the latter.

.. _RFC 3164: https://tools.ietf.org/html/rfc3164
.. _RFC 5424: https://tools.ietf.org/html/rfc5424


:Copyright: 2007-2021 `Jochen Kupperschmidt <https://homework.nwsnet.de/>`_
:License: MIT, see LICENSE for details.


CHANGES.rst
Changelog
=========


Version 0.14
------------

Unreleased

- Updated rtoml to 0.7.1 (from 0.6.1).


Version 0.13
------------

Released 2021-05-07

- Changed log level of message for failed IRC channel join due to bad
  key to "warning". This keeps it visible if one decides logging on
  level "info" is too noisy in production.

- Simplified implementation of ``format_message``. As it is suggested to
  be copied and adjusted by users, this change should make that easier.

- Corrected Weitersager version in ``Dockerfile``.


Version 0.12
------------

Released 2021-05-05

- Switched to using Python's logging mechanism directly, with different
  log levels. Removed function ``util.log``.

- Made the application's log level configurable.

- Introduced an internal queue to decouple syslog message reception from
  message formatting and posting to IRC.

- Added TCP support for syslog message receiving. Adjusted routes
  configuration format to support it.

- Added support for custom IRC commands to send after connecting
  (authentication, cloaking, flood protection, etc.) (idea and essential
  patch provided by Jonas Alexandersson).

- Prevent enabling of a joined IRC channel to receive messages if no
  ports are routed to it.

- Remove leading and trailing newlines from message text to prevent
  additional lines on IRC (patch provided by Jonas Alexandersson).

- Provided and documented an easier way to customize message formatting.


Version 0.11
------------

Released 2021-04-21

- Introduced configuration file. Removed the CLI arguments that have
  been replaced by it.

- Moved configuration of IRC channels to join and ports-to-channels
  routes to configuration file.

- Provided an actual ``syslog2irc`` command.

- Published the package to the Python Package Index to allow installing
  from there.

- Added support for IRC server password.

- Added support for a rate limit for the IRC connection, i.e. the
  maximum number of messages per second to send. This can prevent the
  bot from getting kicked (or even banned) from a channel because of
  flooding.

- Added ``Dockerfile``.

- Added command line option ``--version`` to show syslog2IRC's version.


Version 0.10
------------

Released 2021-04-20

- Removed support for unsupported Python versions 2.7, 3.3, 3.4, and
  3.5.

- Added support for Python 3.7, 3.8, and 3.9.

- Split the single module into several modules inside a namespace
  package and a start script. Refactored a lot.

- Added type hints.

- Expect channels in routing configuration to be in sets, not lists.

- A custom formatter for syslog messages can be specified.

- Made bot properly disconnect from IRC with a quit message on shutdown.

- Removed handler to request shutdown via IRC private message.

- Require blinker version 1.4.

- Require irc version 19.0.1.

- Require syslogmp version 0.4.


Version 0.9.1
-------------

Released 2015-09-09

- Exposed programmatic entry point to pass arguments from Python code
  rather than the command line.

- Introduced signal to indicate an approved message. Decoupled processor
  and announcer.

- Wrapped ``print`` calls in a slightly higher-level, timestamped
  logging function.

- Added and improved tests.

- Created manifest template to package all files relevant for
  distribution.


Version 0.9
-----------

Released 2015-09-08

- Adapted to version 0.2 of syslogmp.

- Spread tests to separate modules.

- Added support to connect to IRC servers via SSL (suggested by Jonas
  Alexandersson).


Version 0.8
-----------

Released 2015-08-10

- Dropped support for Python 3.2.

- The syslog message parser was moved into the new 'syslogmp' package.


Version 0.7
-----------

Released 2014-05-19


Version 0.6
-----------

Released 2014-05-15


Version 0.5
-----------

Released 2013-07-22


Version 0.2–0.4
---------------

unknown


Version 0.1
-----------

Released 2007-04-12


syslog2irc.py
#!/usr/bin/env python
# -*- coding: utf-8 -*-

"""
syslog2IRC
==========

Receive syslog messages via UDP and show them on IRC.

:Copyright: 2007-2015 `Jochen Kupperschmidt <http://homework.nwsnet.de/>`_
:Date: 09-Sep-2015
:License: MIT, see LICENSE for details.
:Version: 0.9.2-dev
"""

from __future__ import print_function
import argparse
from collections import defaultdict, namedtuple
from datetime import datetime
from itertools import chain
try:
    # Python 2.x
    from SocketServer import BaseRequestHandler, ThreadingUDPServer
except ImportError:
    # Python 3.x
    from socketserver import BaseRequestHandler, ThreadingUDPServer
from ssl import wrap_socket as ssl_wrap_socket
import sys
from threading import Thread
from time import sleep

from blinker import signal
import irc
from irc.bot import ServerSpec, SingleServerIRCBot
import syslogmp


DEFAULT_IRC_PORT = ServerSpec('').port

SYSLOG_MESSAGE_TEXT_ENCODING = 'utf-8'


# A note on threads (implementation detail):
#
# This tool uses threads. Besides the main thread, there are two
# additional threads: one for the syslog message receiver and one for
# the IRC bot. Both are configured to be daemon threads.
#
# A Python application exits if no more non-daemon threads are running.
#
# In order to exit syslog2IRC when shutdown is requested on IRC, the IRC
# bot will call `die()`, which will join the IRC bot thread. The main
# thread and the (daemonized) syslog message receiver thread remain.
#
# Additionally, a dedicated signal is sent that sets a flag that causes
# the main loop to stop. As the syslog message receiver thread is the
# only one left, but runs as a daemon, the application exits.
#
# The STDOUT announcer, on the other hand, does not run in a thread. The
# user has to manually interrupt the application to exit.
#
# For details, see the documentation on the `threading` module that is
# part of Python's standard library.


# -------------------------------------------------------------------- #
# signals


syslog_message_received = signal('syslog-message-received')
irc_channel_joined = signal('irc-channel-joined')
message_approved = signal('message-approved')
shutdown_requested = signal('shutdown-requested')


# -------------------------------------------------------------------- #
# syslog


def format_syslog_message(message):
    """Format a syslog message to be displayed."""
    def _generate():
        if message.timestamp is not None:
            timestamp_format = '%Y-%m-%d %H:%M:%S'
            formatted_timestamp = message.timestamp.strftime(
                timestamp_format)
            yield '[{}] '.format(formatted_timestamp)

        if message.hostname is not None:
            yield '({}) '.format(message.hostname)

        severity_name = message.severity.name
        # Important: The message text is a byte string.
        message_text = message.message.decode(SYSLOG_MESSAGE_TEXT_ENCODING)
        yield '[{}]: {}'.format(severity_name, message_text)

    return ''.join(_generate())


class SyslogRequestHandler(BaseRequestHandler):
    """Handler for syslog messages."""

    def handle(self):
        try:
            data = self.request[0]
            message = syslogmp.parse(data)
        except ValueError:
            log('Invalid message received from {}:{:d}.',
                *self.client_address)
            return

        port = self.server.get_port()
        syslog_message_received.send(port,
            source_address=self.client_address, message=message)


class SyslogReceiveServer(ThreadingUDPServer):
    """UDP server that waits for syslog messages."""

    def __init__(self, port):
        ThreadingUDPServer.__init__(self, ('', port),
            SyslogRequestHandler)

    @classmethod
    def start(cls, port):
        """Start in a separate thread."""
        try:
            receiver = cls(port)
        except Exception as e:
            sys.stderr.write('Error {0.errno:d}: {0.strerror}\n'
                .format(e))
            sys.stderr.write(
                'Probably no permission to open port {}. '
                'Try to specify a port number above 1,024 (or even '
                '4,096) and up to 65,535.\n'.format(port))
            sys.exit(1)

        thread_name = '{}-port{:d}'.format(cls.__name__, port)
        start_thread(receiver.serve_forever, thread_name)

    def get_port(self):
        return self.server_address[1]


def start_syslog_message_receivers(ports):
    """Start one syslog message receiving server for each port."""
    for port in ports:
        SyslogReceiveServer.start(port)


# -------------------------------------------------------------------- #
# IRC


class IrcChannel(namedtuple('IrcChannel', 'name password')):
    """An IRC channel with optional password."""

    def __new__(cls, name, password=None):
        return super(IrcChannel, cls).__new__(cls, name, password)


class IrcBot(SingleServerIRCBot):
    """An IRC bot to forward syslog messages to IRC channels."""

    def __init__(self, server_spec, nickname, realname, channels, ssl=False):
        log('Connecting to IRC server {0.host}:{0.port:d} ...', server_spec)

        connect_params = {}
        if ssl:
            ssl_factory = irc.connection.Factory(wrapper=ssl_wrap_socket)
            connect_params['connect_factory'] = ssl_factory

        SingleServerIRCBot.__init__(self, [server_spec], nickname,
            realname, **connect_params)

        # Note: `self.channels` already exists in super class.
        self.channels_to_join = channels

    def get_version(self):
        """Return this on CTCP VERSION requests."""
        return 'syslog2IRC'

    def on_welcome(self, conn, event):
        """Join channels after connect."""
        log('Connected to {}:{:d}.', *conn.socket.getpeername())

        channel_names = sorted(c.name for c in self.channels_to_join)
        log('Channels to join: {}', ', '.join(channel_names))

        for channel in self.channels_to_join:
            conn.join(channel.name, channel.password or '')

    def on_nicknameinuse(self, conn, event):
        """Choose another nickname if conflicting."""
        self._nickname += '_'
        conn.nick(self._nickname)

    def on_join(self, conn, event):
        """Successfully joined channel."""
        joined_nick = event.source.nick
        channel_name = event.target

        if joined_nick == self._nickname:
            log('Joined IRC channel: {}', channel_name)
            irc_channel_joined.send(channel=channel_name)

    def on_badchannelkey(self, conn, event):
        """Channel could not be joined due to wrong password."""
        channel_name = event.arguments[0]
        log('Cannot join channel {} (bad key).', channel_name)

    def on_privmsg(self, conn, event):
        """React on private messages.

        Shut down, for example.
        """
        nickmask = event.source
        text = event.arguments[0]

        if message == 'shutdown!':
            self.shutdown(nickmask)

    def shutdown(self, nickmask):
        """Shut the bot down."""
        log('Shutdown requested by {}.', nickmask)
        shutdown_requested.send()
        self.die('Shutting down.')  # Joins IRC bot thread.

    def say(self, channel, message):
        """Say message on channel."""
        self.connection.privmsg(channel, message)


# -------------------------------------------------------------------- #
# announcing


class IrcAnnouncer(object):
    """Announce syslog messages on IRC."""

    def __init__(self, server, nickname, realname, channels, ssl=False):
        self.bot = IrcBot(server, nickname, realname, channels, ssl=ssl)

    def start(self):
        start_thread(self.bot.start, 'IrcAnnouncer')

    def announce(self, sender, channel_name=None, text=None):
        self.bot.say(channel_name, text)


class StdoutAnnouncer(object):
    """Announce syslog messages on STDOUT."""

    def start(self):
        pass

    def announce(self, sender, channel_name=None, text=None):
        log('{}> {}', channel_name, text)


def create_announcer(irc_server, irc_nickname, irc_realname,
                     irc_channels, **options):
    """Create and return an announcer according to the configuration."""
    if not irc_server:
        log('No IRC server specified; will write to STDOUT instead.')
        return StdoutAnnouncer()

    return IrcAnnouncer(irc_server, irc_nickname, irc_realname,
                        irc_channels, **options)


# -------------------------------------------------------------------- #
# utilities


def log(message, *args, **kwargs):
    """Log the message with a timestamp."""
    timestamp = datetime.now().strftime('%Y-%m-%d %H:%M:%S')
    print(timestamp, message.format(*args, **kwargs))


def start_thread(target, name):
    """Create, configure, and start a new thread."""
    t = Thread(target=target, name=name)
    t.daemon = True
    t.start()


# -------------------------------------------------------------------- #


class Runner(object):

    def __init__(self):
        self.shutdown = False

    def request_shutdown(self, sender):
        self.shutdown = True

    def run(self, seconds_to_sleep=0.5):
        """Run the main loop until shutdown is requested."""
        while not self.shutdown:
            sleep(seconds_to_sleep)

        log('Shutting down ...')


class Processor(Runner):

    def __init__(self, channel_names_to_ports):
        super(Processor, self).__init__()

        self.channel_names_to_ports = channel_names_to_ports
        self.ports_to_channel_names = defaultdict(set)

    def connect_to_signals(self):
        irc_channel_joined.connect(self.enable_channel)
        shutdown_requested.connect(self.request_shutdown)
        syslog_message_received.connect(self.handle_syslog_message)

    def enable_channel(self, sender, channel=None):
        ports = self.channel_names_to_ports[channel]

        log('Enabled forwarding to channel {} from ports {}.',
            channel, ports)

        for port in ports:
            self.ports_to_channel_names[port].add(channel)

    def handle_syslog_message(self, port, source_address=None,
            message=None):
        """Log and process an incoming syslog message."""
        source = '{0[0]}:{0[1]:d}'.format(source_address)

        log('Received message from {} on port {:d} -> {}',
            source, port, format_message_for_log(message))

        formatted_message = format_syslog_message(message)
        irc_message = '{} {}'.format(source, formatted_message)
        for channel_name in self.ports_to_channel_names[port]:
            message_approved.send(channel_name=channel_name,
                                  text=irc_message)


def format_message_for_log(message):
    facility_name = message.facility.name
    severity_name = message.severity.name
    timestamp_str = message.timestamp.isoformat()
    hostname = message.hostname

    return 'facility={}, severity={}, timestamp={}, hostname={}, message={}' \
           .format(facility_name, severity_name, timestamp_str, hostname,
                   message.message)


# -------------------------------------------------------------------- #
# command line argument parsing


def parse_args(args=None):
    """Parse command line arguments."""
    parser = create_arg_parser()
    return parser.parse_args(args)


def create_arg_parser():
    """Prepare the command line arguments parser."""
    parser = argparse.ArgumentParser()

    parser.add_argument('--irc-nickname',
        dest='irc_nickname',
        default='syslog',
        help='the IRC nickname the bot should use',
        metavar='NICKNAME')

    parser.add_argument('--irc-realname',
        dest='irc_realname',
        default='syslog2IRC',
        help='the IRC realname the bot should use',
        metavar='REALNAME')

    parser.add_argument('--irc-server',
        dest='irc_server',
        type=parse_irc_server_arg,
        help='IRC server (host and, optionally, port) to connect to'
            + ' [e.g. "irc.example.com" or "irc.example.com:6669";'
            + ' default port: {:d}]'.format(DEFAULT_IRC_PORT),
        metavar='SERVER')

    parser.add_argument('--irc-server-ssl',
        dest='irc_server_ssl',
        action='store_true',
        help='use SSL to connect to the IRC server')

    return parser


def parse_irc_server_arg(value):
    """Parse a hostname with optional port."""
    fragments = value.split(':', 1)
    if len(fragments) > 1:
        fragments[1] = int(fragments[1])
    return ServerSpec(*fragments)


# -------------------------------------------------------------------- #

def map_channel_names_to_ports(routes):
    channel_names_to_ports = defaultdict(set)
    for port, channels in routes.items():
        for channel in channels:
            channel_names_to_ports[channel.name].add(port)
    return channel_names_to_ports


def start(irc_server, irc_nickname, irc_realname, routes, **options):
    """Start the IRC bot and the syslog listen server."""
    irc_channels = frozenset(chain(*routes.values()))
    channel_names_to_ports = map_channel_names_to_ports(routes)
    ports = routes.keys()

    announcer = create_announcer(irc_server, irc_nickname, irc_realname,
                                 irc_channels, **options)
    message_approved.connect(announcer.announce)

    processor = Processor(channel_names_to_ports)

    # Up to this point, no signals must have been sent.
    processor.connect_to_signals()

    # Signals are allowed be sent from here on.

    start_syslog_message_receivers(ports)
    announcer.start()

    if not irc_server:
        # Fake channel joins.
        for channel in channel_names_to_ports.keys():
            irc_channel_joined.send(channel=channel)

    processor.run()


def start_with_args(routes):
    """Start the IRC bot and the syslog listen server.

    All arguments (except for routes) are read from the command line.
    """
    args = parse_args()

    start(args.irc_server, args.irc_nickname, args.irc_realname, routes,
          ssl=args.irc_server_ssl)


if __name__ == '__main__':
    # IRC channels to join
    irc_channel1 = IrcChannel('#examplechannel1')
    irc_channel2 = IrcChannel('#examplechannel2', password='zePassword')

    # routing for syslog messages from the ports on which they are
    # received to the IRC channels they should be announced on
    routes = {
          514: [irc_channel1, irc_channel2],
        55514: [irc_channel2],
    }

    start_with_args(routes)
